Add build configuration to generate runtime packs for iOS

steveisok opened the pull request:
This adds support for generating iOS runtime packs, which is intended to be consumed by the Xamarin iOS SDK.

Commit:
Add build configuration to generate runtime packs for iOS

This adds support for generating iOS runtime packs, which
is intended to be consumed by the Xamarin iOS SDK.

marek-safar commented on eng/Subsets.props:
```diff
@@ -218,15 +220,30 @@
+    <PkgprojProjectToBuild  
+        Include="$(InstallerProjectRoot)pkg\projects\netcoreapp\pkg\Microsoft.NETCore.App.Runtime.pkgproj" 
+        SignPhase="MsiFiles" BuildInParallel="false" />
+    <PkgprojProjectToBuild  
```
Could you add comment what legacy and workaround is about?

akoeplinger replied:
Those are pre-existing projects/directories but yeah it'd be good to find out what they're used for and whether we can simplify.

steveisok replied:
We may not need it if I can handle it at the pkgproj level.

@dagood Do you know the history?

dagood replied:
Legacy produces the old nupkgs that formed NETCoreApp in 2.1. Workaround produces a package to work around some issues that were found in 3.0 SDK development. https://github.com/dotnet/core-setup/pull/6096, https://github.com/dotnet/runtime/issues/3528.

But yes, I do not think they should be called out explicitly in this file. Adding a readme to the project dir or inside the project files would be the way to go IMO.

dagood replied:
> whether we can simplify.

Probably. However, the Host tests (`src/installer/test/Assets`) currently still depend on the legacy packages. The legacy packages were also used for testing by downstream repos even after 3.0, and may still be.

akoeplinger commented on eng/Subsets.props:
```diff
@@ -58,8 +58,10 @@
+    <DefaultSubsetCategories Condition="'$(TargetOS)' == 'iOS'">libraries-installer-mono</DefaultSubsetCategories>
+    <DefaultSubsetCategories Condition="'$(TargetOS)' == 'Android'">libraries-mono</DefaultSubsetCategories>
     <DefaultInstallerSubsets>corehost-managed-depproj-pkgproj-bundle-installers-test</DefaultInstallerSubsets>
+    <DefaultInstallerSubsets Condition="'$(TargetOS)' == 'iOS'">managed-depproj-pkgproj-bundle-test</DefaultInstallerSubsets>
```
why do we need anything other than the `depproj` and `pkgproj ` subsets?

mdh1418 commented on eng/liveBuilds.targets:
```diff
@@ -122,8 +127,14 @@
+    <Error Condition="'@(CoreCLRFiles)' == ''" Text="The $(RuntimeFlavor) subset category must be built before building this project." />
   </Target>
 
   <Target Name="ResolveMonoFilesFromLocalBuild">
```
Do we still need this target?

mdh1418 replied:
If we also change `ResolveMonoFilesFromLocalBuild` to `ResolveRuntimeFilesFromLocalBuild` or even a more concise way to combine the two in `eng/referenceFromRuntime/targets`?

steveisok replied:
I wasn't sure how big of a bite to take.  We would take away `ResolveMonoFilesFromLocalBuild` and probably want to change `<CoreCLRFiles>` too

dagood commented on eng/Subsets.props:
```diff
@@ -218,15 +220,30 @@
     <ProjectToBuild Include="@(PkgprojProjectToBuild)" />
   </ItemGroup>
 
+  <ItemGroup Condition="$(_subsetCategory.Contains('installer')) and $(_subset.Contains('-pkgproj-')) and '$(TargetOS)' == 'iOS'">    
```
My initial impression is that I don't like seeing this complexity in the subsets system. It seems more maintainable and readable (discoverable) to me for the pkgproj/depproj/etc. to be able to do the right thing when encountering this new TargetOS.

dagood replied:
For example, pkgproj Build has functionality to allow a complete skip:

https://github.com/dotnet/arcade/blob/7c5a68c1aeafec3a9a1b63d1d2c3b197964cc1b7/src/Microsoft.DotNet.Build.Tasks.SharedFramework.Sdk/targets/framework.packaging.targets#L237-L239

A potential way is to manipulate `BuildOnUnknownPlatforms` to trigger a skip based on RID not being present:

https://github.com/dotnet/runtime/blob/1d018ba7e9bfcbc8eb1f56fd14532e7fcfc480c1/src/installer/pkg/projects/Directory.Build.props#L117-L119

https://github.com/dotnet/arcade/blob/866b2acd5ffc3c2031c102f6415fd7c6a1a370d5/src/Microsoft.DotNet.Build.Tasks.SharedFramework.Sdk/targets/packaging-tools.targets#L167-L168

steveisok replied:
So, I skipped at the pkgproj level.  It worked, but it still generated an empty nupkg for each project.  It's unclear if I can actually skip that step.  Do you know?

dagood replied:
Yes, disabling `BuildOnUnknownPlatforms` prevents the WindowsDesktop packages from building where they shouldn't in Core-Setup release/3.1. Wouldn't expect to see empty nupkgs in any situation I know about.

dagood commented on eng/liveBuilds.targets:
```diff
@@ -78,10 +78,11 @@
 
-  <Target Name="ResolveCoreCLRFilesFromLocalBuild">
-    <Error Condition="!Exists('$(CoreCLRArtifactsPath)')" Text="The CoreCLR artifacts path does not exist '$(CoreCLRArtifactsPath)'. The CoreCLR subset category must be built before building this project." />
+  <Target Name="ResolveRuntimeFilesFromLocalBuild">
```
Should we say `VM` or something better, to avoid being ambiguous between "CoreCLR vs. Mono" and "the Runtime repository"?

Commit:
Take away unnecessary subsets / fix syntax error

dagood commented on eng/pipelines/installer/jobs/base-job.yml:
```diff
@@ -186,7 +186,6 @@ jobs:
       value: >-
         --restore --build --ci --test
         /p:CrossBuild=${{ ne(parameters.crossrootfsDir, '') }}
-        /p:DisableCrossgen=$(_DisableCrossgen)
```
Clean up unused variable earlier in file?

dagood commented on src/installer/Directory.Build.props:
```diff
@@ -129,7 +130,7 @@
     <DisableCrossgen>false</DisableCrossgen>
     <!-- Disable cross-gen on FreeBSD for now. This can be revisited when we have full support. -->
-    <DisableCrossgen Condition="'$(TargetOS)'=='FreeBSD'">true</DisableCrossgen>
+    <DisableCrossgen Condition="'$(TargetOS)'=='FreeBSD' or '$(RuntimeFlavor)' == 'Mono'">true</DisableCrossgen>
```
Put on separate line to simplify FreeBSD fix later, and clarify the target of the comment?

dagood commented on src/installer/Directory.Build.props:
```diff
@@ -147,10 +148,10 @@
+    <InstallerStartSuffix Condition="'$(TargetOS)' == 'OSX' or '$(TargetOS)' == 'iOS'">internal</InstallerStartSuffix>
+    <SharedHostInstallerStart Condition="'$(TargetOS)' == 'OSX' or '$(TargetOS)' == 'iOS'">$(SharedHostInstallerStart)$(InstallerStartSuffix)-</SharedHostInstallerStart>
+    <HostFxrInstallerStart Condition="'$(TargetOS)' == 'OSX' or '$(TargetOS)' == 'iOS'">$(HostFxrInstallerStart)$(InstallerStartSuffix)-</HostFxrInstallerStart>
+    <SharedFrameworkInstallerStart Condition="'$(TargetOS)' == 'OSX' or '$(TargetOS)' == 'iOS'">$(SharedFrameworkInstallerStart)$(InstallerStartSuffix)-</SharedFrameworkInstallerStart>
```
We don't plan to produce shared framework installers for iOS, do we? (Add to comment about this? The change itself seems harmless either way but it stands out and could be confusing.)

steveisok replied:
No, I'll take it away.

dagood commented on src/installer/Directory.Build.props:
```diff
@@ -276,7 +284,7 @@
     <CompressedFileExtension Condition="'$(TargetOS)' != 'Windows_NT'">.tar.gz</CompressedFileExtension>
     <InstallerExtension Condition="'$(TargetOS)' == 'Windows_NT'">.msi</InstallerExtension>
-    <InstallerExtension Condition="'$(TargetOS)' == 'OSX'">.pkg</InstallerExtension>
+    <InstallerExtension Condition="'$(TargetOS)' == 'OSX' or '$(TargetOS)' == 'iOS'">.pkg</InstallerExtension>
```
(Similar as above) Are we really producing a sharedfx pkg for iOS?

steveisok replied:
No, I was taking a scorched earth approach and can take it away.

dagood commented on src/installer/pkg/packaging/osx/package.targets:
```diff
@@ -5,13 +5,13 @@
 
   <Target Name="InitPkg"
-          Condition="'$(TargetOS)' == 'OSX'">
+          Condition="'$(TargetOS)' == 'OSX' or '$(TargetOS') == 'iOS'">
```
Typo, quote misplaced.

dagood commented on src/installer/pkg/projects/netcoreapp/pkg/Microsoft.NETCore.App.Runtime.pkgproj:
```diff
@@ -4,7 +4,7 @@
-  <Import Project="..\..\Microsoft.NETCore.DotNetHostPolicy\Microsoft.NETCore.DotNetHostPolicy.props" />
-  <Import Project="..\..\Microsoft.NETCore.DotNetHostResolver\Microsoft.NETCore.DotNetHostResolver.props" />
+  <Import Condition="'$(TargetOS)' != 'iOS'" Project="..\..\Microsoft.NETCore.DotNetHostPolicy\Microsoft.NETCore.DotNetHostPolicy.props" />
+  <Import Condition="'$(TargetOS)' != 'iOS'" Project="..\..\Microsoft.NETCore.DotNetHostResolver\Microsoft.NETCore.DotNetHostResolver.props" />
```
Is iOS really what matters here? I would expect Mono (*if* we don't want to include the host resolver or host policy libraries with it).

steveisok replied:
I made it iOS here because mono should also be able to generate the same things as coreclr.  Generating just runtime packs are only going to apply to iOS and Android.

dagood replied:
I think it'd be nice to make this clear by adding a property and clarifying comment (that doesn't interfere with existing comment):

```xml
<!--
  Don't include host resolver or host policy in iOS or Android runtime packs, because there is no
  shared framework on those platforms.
-->
<PropertyGroup>
  <ExcludeHostAssets Condition="'$(TargetOS)' == 'iOS'">true</ExcludeHostAssets>
  <ExcludeHostAssets Condition="'$(TargetOS)' == 'Android'">true</ExcludeHostAssets>
</PropertyGroup>

<!--
  Include HostPolicy and HostResolver based on legacy packages. This can be collapsed once the
  legacy packages are no longer used.
-->
<Import Condition="'$(ExcludeHostAssets)' != 'true'" Project="..\..\Microsoft.NETCore.DotNetHostPolicy\Microsoft.NETCore.DotNetHostPolicy.props" />
<Import Condition="'$(ExcludeHostAssets)' != 'true'" Project="..\..\Microsoft.NETCore.DotNetHostResolver\Microsoft.NETCore.DotNetHostResolver.props" />
```

jaredpar commented on eng/referenceFromRuntime.targets:
```diff
@@ -45,7 +45,7 @@
           Condition="'$(RuntimeFlavor)' != 'Mono'"
           Returns="@(CoreCLRFiles)"
-          DependsOnTargets="ResolveCoreCLRFilesFromLocalBuild" />
+          DependsOnTargets="ResolveRuntimeFilesFromLocalBuild" />
```
Why did you change this target but not the `GetFilesFromMono` below? Those targets should mirror each other in this respect.

steveisok replied:
I didn't want to change too much until after I got something working.  Now that I look again, I don't think it's a big deal to alter.

Commit:
Addressed some feedback

dagood commented on eng/liveBuilds.targets:
```diff
@@ -78,10 +78,11 @@
 
-  <Target Name="ResolveCoreCLRFilesFromLocalBuild">
-    <Error Condition="!Exists('$(CoreCLRArtifactsPath)')" Text="The CoreCLR artifacts path does not exist '$(CoreCLRArtifactsPath)'. The CoreCLR subset category must be built before building this project." />
+  <Target Name="ResolveVMFilesFromLocalBuild">
```
I threw this suggestion out there without much context, so I want to highlight it for other reviewers... "runtime" as CoreCLR/Mono vs. "runtime" as the marketing name of the shared framework has been ambiguous for a while, and at some point early in Mono discussions "VM" came up as a possible less ambiguous name for the CoreCLR/Mono grouping. (The implications of "VM" might actually make it a straight-up incorrect suggestion here, I seem to remember some issue like that with the name, e.g. that aot tooling is included so it doesn't make sense.)

The code has already built up concepts like `RuntimeFlavor` which use the first definition, so maybe it's too late and doesn't matter enough to really change--context clues may be enough.

steveisok replied:
For me, either are fine.  I honestly thought runtime fit better, but I'll go with whatever the group decides.

jaredpar replied:
Small preference for `ResolveRuntimeFilesFromLocalBuild` and `RuntimeFiles` because I feel like it has consistency with other property names / item groups we use today. Don't feel particularly strongly about it though.

steveisok replied:
I prefer RuntimeFiles.  Who gets to decide?

dagood replied:
I put the dotnet/runtime-infrastructure team as a reviewer to discuss/reject the name. I believe the general model is that it's up to your judgment unless repo or area owners care otherwise. So far it sounds like consistency in property naming + not overhauling properties during this PR outweighs the ambiguity, and I don't disagree with that.

Commit:
Attempt to add iOS installer lanes

Commit:
Trying a single platform to see if it'll work

Commit:
I guess there's one already?

Commit:
Adding ios installer ci back in.  It's gonna fail

mdh1418 commented on eng/liveBuilds.targets:
```diff
@@ -78,10 +78,11 @@
+    <Error Condition="!Exists('$(MonoArtifactsPath)') and '$(RuntimeFlavor)' == 'Mono'" Text="The Mono artifacts path does not exist '$(MonoArtifactsPath)'. The Mono subset category must be built before building this project." />
 
-    <PropertyGroup>
+    <PropertyGroup Condition="'$(RuntimeFlavor)' == 'CoreCLR'">
```
Would it be clearer if we group the Property/Item groups by their conditions?
```
<Choose>
   <When Condition="'$(RuntimeFlavor)' == 'CoreCLR'">
      <PropertyGroup></PropertyGroup>
      <ItemGroup></ItemGroup>
   </When>
   <When Condition="'$(RuntimeFlavor)' == 'Mono'">
      <PropertyGroup></PropertyGroup>
      <ItemGroup></ItemGroup>
   </When>
</Choose>
```

akoeplinger replied:
It looks like `<Choose>` only works at the project level, not inside a Target: https://github.com/Microsoft/msbuild/issues/2056

mdh1418 commented on src/installer/Directory.Build.props:
```diff
@@ -128,8 +129,9 @@
+    <DisableCrossgen Condition="'$(RuntimeFlavor)' == 'Mono'">true</DisableCrossgen>
     <!-- Disable cross-gen on FreeBSD for now. This can be revisited when we have full support. -->
-    <DisableCrossgen Condition="'$(TargetOS)'=='FreeBSD'">true</DisableCrossgen>
+    <DisableCrossgen Condition="'$(TargetOS)'=='FreeBSD'">true</DisableCrossgen> 
```
remove extra space

mdh1418 commented on src/installer/Directory.Build.props:
```diff
@@ -93,6 +93,7 @@
   <PropertyGroup Condition="'$(PortableBuild)' == 'true'">
     <OutputRid Condition="'$(TargetOS)' == 'Windows_NT'">win-$(TargetArchitecture)</OutputRid>
     <OutputRid Condition="'$(TargetOS)' == 'OSX'">osx-$(TargetArchitecture)</OutputRid>
+    <OutputRid Condition="'$(TargetOS)' == 'iOS'">iOS-$(TargetArchitecture)</OutputRid>
```
lowercase for consistency? Or is the iOS Rid prefix already used elsewhere?

safern commented on eng/pipelines/runtime.yml:
```diff
@@ -481,6 +481,18 @@ jobs:
       liveRuntimeBuildConfig: release
       liveLibrariesBuildConfig: ${{ variables.debugOnPrReleaseOnRolling }}
 
+- template: /eng/pipelines/installer/installer-matrix.yml
```
This is only adding the legs to CI, if we want the runtime packages to be published, this needs to be added to runtime-official.yml as well.

Commit:
Tweaked platform matrix for ios_arm & ios_arm64 so that they do not repeat iOS_x64. This was causing the installer leg creation to fail

Commit:
Added iOS as part of the base job check

Commit:
Added os and archType to the ios installer args

Commit:
yml syntax :-(

Commit:
Yml syntax

safern commented on eng/pipelines/installer/jobs/base-job.yml:
```diff
@@ -112,7 +112,7 @@ jobs:
         $(MsbuildSigningArguments)
 
-  - ${{ if eq(parameters.osGroup, 'OSX') }}:
+  - ${{ if or(eq(parameters.osGroup, 'OSX'), eq(parameters.osGroup, 'iOS')) }}:
```
You can simplify this condition by using `in` conditional expression.

```yml
${{ if in(parameters.osGroup, 'OSX', 'iOS') }}:
```

That way you don't need to do an or of eq statements.

jaredpar commented on src/libraries/restore/runtime/runtime.depproj:
```diff
@@ -74,27 +74,27 @@
   </Target>
 
   <Target Name="OverrideRuntimeCoreCLR"
```
If we're unifying the `Resolve*FilesFromLocalBuild` should we unify the `OverrideRuntime*` target as well?

steveisok replied:
I'd prefer doing that in another PR.

safern commented on eng/liveBuilds.targets:
```diff
@@ -91,26 +92,30 @@
+    </PropertyGroup>
+
+    <ItemGroup Condition="'$(RuntimeFlavor)' == 'CoreCLR'">
+      <VMFiles Include="$(CoreCLRSharedFrameworkDir)*.*" />
```
`RuntimeFiles` to be consistent with properties like `runtimeConfiguration` and `RuntimeFlavor`?

Commit:
Pushed skipping unnecessary pkgproj into installer/pkg area

Added cross and include into the native part of the runtime pack

Addressed feedback

Commit:
Adjusted skip props for host pkgproj

Commit:
More feedback

steveisok commented on src/installer/pkg/projects/Microsoft.NETCore.DotNetHostPolicy/Microsoft.NETCore.DotNetHostPolicy.pkgproj:
```diff
@@ -8,6 +8,12 @@
   </ItemGroup>
 
-  <Import Project="$(MSBuildProjectName).props" />
+  <Import Project="$(MSBuildProjectName).props" Condition="'$(TargetOS)' != 'iOS'" />
```
From what I can tell, `$(PackageTargetRuntime)` has a value, so `$(TargetOS)` or some kind of other check seems to be needed.

dagood replied:
Maybe I'm misinterpreting your comment, but yes, that's why I suggested checking for existence of the file. 😄 That would handle the generic case.

dagood commented on src/installer/pkg/projects/Directory.Build.targets:
```diff
@@ -14,6 +14,17 @@
+    in in order to skip building. 
+  -->
+  <Target Name="MobileSkipBuildProps" BeforeTargets="GetSkipBuildProps" 
+    Condition="'$(SkipBuildOnRuntimePackOnlyOS)' and '$(TargetOS)' == 'iOS'">
```
(Nit) I find this style of wrapping significantly harder to read quickly vs. chopping or alignment. `BeforeTargets` is hard to see, and `Condition` being left of `Name` is unusual. This is a thing that maybe you can get used to, but thought I'd mention it anyway while I'm here and because it's in src/installer.

safern replied:
Nit, we usually not use boolean conditions based from properties, we usually do direct string comparison ourselves, because boolean condition, forces you to set it to either false or true in all code paths. If we made this a boolean, the default could be false, so sometime like `'$(SkipBuildOnRuntimePackOnlyOS)' != 'true'` then that way you only set it to true when you want to disable this target but not set it to false anywhere.

Commit:
More feedback

Commit:
More feedback

Commit:
yml syntax in runtime official

akoeplinger commented on src/installer/pkg/projects/Directory.Build.targets:
```diff
@@ -14,6 +14,18 @@
 
+  <!--
+    For mobile targets, we do not need the host, so allow depproj's to opt 
+    in in order to skip building. 
```
opt in -> opt out

Commit:
Last bit of nits

Commit:
Disabled useOfficialAllConfigurations and buildFullPlatformManifest on the official build.
